### DISCUSSION TITLE
Milvus query-nodes are sensitive to K8S disruptions

### DISCUSSION THREAD

**Discussion by mkotsalainen** [milvus-io/milvus#38283] [Q&A and General discussion]
Hi! We're trying to run our system as cheaply as possible and that means that we run only replica of each collection, we run on spot nodes and we've got [Karpenter consolidation](https://karpenter.sh/v0.32/concepts/disruption/#consolidation) enabled. We see that this causes a couple of minutes downtime when a query node is being rescheduled.

Is there something we can do to improve this? We often get a couple of minutes warning that the disruption will occur - so in theory it should be enough for the new query node to be scheduled and load its data from S3.

**Comment by mkotsalainen** [comment]
I see now that this has been brought up previously: https://github.com/milvus-io/milvus/issues/20932 It would be a nice feature to add to make Milvus cheaper to run.

**Comment by LoveEachDay** [comment]
@mkotsalainen So it's ok to only query online segments for your scenario? By default, milvus only serving search only after all the segments are loaded for specified collection.

When a qn rescheduled, the segments will be reloaded into other querynodes, which will cost disruptions.

**Reply by mkotsalainen** [reply]
No - I guess what I'm asking for is some way to get Milvus to preemptively spin up a new qn before terminating the old. Since we know a few minutes in advance that the pod will be terminated it would be nice to use that to our advantage to start provisioning the new qn before the old is terminated.

Right now I think my only option is to put the qns on more expensive static machines.

**Reply by xiaofan-luan** [reply]
@LoveEachDay Do we have an API in K8s operator to graceful shutdown once server?
Actaully I believe it should be 
1. bring up a new pod
2. kill one pod
3. stop the pod
